domain/schema/model.go

```diff
@@ -558,11 +599,92 @@ INSERT INTO storage_kind VALUES
     (0, 'block', 'Allows for the creation of raw storage volumes'), 
     (1, 'filesystem', 'Provides a hierarchical file storage system');
 
+-- This table stores storage constraint values for each named storage item
+-- defined by the application's current charm. If the charm is updated, then
+-- so too will be the rows in this table to reflect the current charm's
```

Comment:
This doesn't really make sense to me. An application has one charm.

A _unit_ might be on an old version of the charm compared to the application, but today that is indicated by charm URL. Given that a charm URL uniquely identifies a charm today, I suppose we will probably have a new charm record (new UUID) that we change the application to, then the units progressively.

To me it makes more sense to store a set of constraints declared by the charm against the charm itself, have this table keyed just on the application, and build the constraints hierarchically.

Reply:
Storage constraints are tied to an application not a charm. The charm does not itself declare storage constraints. The charm defines the allowed set of storage names it supports, in the charm_storage table. We create storage constraints for each application which uses that charm - each app storage constraint is for a charm storage name from the charm storage table. Each application will have a different set of constraints.

Reply:
Isn't minimum size against the charm storage effectively a default constraint?

Reply:
Minimum size can be considered like that but it's not what "storage constraints" end up being. Actually, we really should call the juju model artefact "storage directive" and the table becomes "storage_directive". This is the correct terminology that we should be using to refer to this aspect. We talk about deployment with "--storage" directives. The allowed storage directive is informed / constrained by the rules in charm metadata (min/max count aka range, min size) but ultimately specified by the user with model / storage kind fallbacks. So the charm metadata constrains "pgdata" storage to be min 10GB and between 1 and 3 attached volumes. The user deploys postgresql with storage directive "--storage ebs,20GB,2". This falls within the bounds of the charm specified constraints so that means every unit of postgresql is realised with that exact storage requirement. Note that the charm does not specify the provider type (eg ebs), just the kind. The exact type is supplied as part of the storage directive by the user, so long as it matches the storage kind it's allowed.

So TL;DR; we should call this table "storage_directive". Then I hope the modelling of t makes more sense.